cmd/web-handlers.go

```diff
@@ -330,8 +330,43 @@ func (web *webAPIHandlers) ListBuckets(r *http.Request, args *WebGenericArgs, re
 	} else {
 		buckets, err := listBuckets(ctx)
 		if err != nil {
+			if u, parseerr := url.Parse(r.Referer()); parseerr == nil {
+				bucket, prefix := path2BucketObjectWithBasePath(minioReservedBucketPath, u.Path)
+				if bucket != "" && bucket != u.Path {
+					readable := globalIAMSys.IsAllowed(iampolicy.Args{
+						AccountName:     claims.AccessKey,
+						Action:          iampolicy.ListBucketAction,
+						BucketName:      bucket,
+						ObjectName:      prefix + SlashSeparator,
+						ConditionValues: getConditionValues(r, "", claims.AccessKey, claims.Map()),
+						IsOwner:         owner,
+						Claims:          claims.Map(),
+					})
+					if !strings.HasSuffix(prefix, slashSeparator) {
+						prefix += slashSeparator
+					}
+					writable := globalIAMSys.IsAllowed(iampolicy.Args{
```

Comment:
Why do you need putobject permission?

Reply:
PutObject is how 'can write' access is deduced. It's standard throughout web-handlers.go. No read and write access implies we should deny access.